### ISSUE TITLE
How to use `configurator` to fetch `pkg-config` for the cross-compiling env?

### ISSUE THREAD

**Issue by toots** [ocaml/dune#2572]
`ocaml-shine` has now been converted to use `dune` as build system!

However, when cross-compiling for `windows`, the [discover](https://github.com/savonet/ocaml-shine/blob/master/src/config/discover.ml) program used to fetch `pkg-config` information is compiled as a native binary and returns `pkg-info` related to the build host and not the target cross-compilation.

In `configure` world, this is handled by passing a host triplet and expecting the binary to use to also be prefixed with the same host triple, i.e. `x86_64-w64-mingw32-pkg-config` and etc.

Could it be possible to make this work with `configurator`?

**Comment by toots** [comment]
Just wanted to bring this up again. When cross-compiling with dune, I now set `PKG_CONFIG_PATH` to the cross-compiler's path to `pkg-config` files. However, this is tricky. Since `dune` can implicitly also build binaries for the native environment if e.g. they need to be run locally, this will, in turn, trigger the `discover` script.

However, forcing `PKG_CONFIG_PATH` means that the `discover` script only sees the cross-compiling pkg-config info so we cannot configure the build to be able to build both for the native and target host using `pkg-config`.

In an ideal work, it should be possible to use either the `context` variable or, possibly, the `toolchain` variable (#3949) to write a `discover` module using `dune-configurator` that would pick the correct `pkg-config` binary and not have to worry about internals that are already pre-configured for each binary.

**Comment by jeremiedimino** [comment]
We just discussed this during the dune dev meeting. It seems like adding a `pkg-config` variable to the ocamlfind toolchain definition could help here. Then configurator would call this `pkg-config` binary. WDYT?

**Comment by toots** [comment]
I think that this would work great!

**Comment by anmonteiro** [comment]
@toots I'm fixing this in https://github.com/ocaml/dune/pull/7469 (respecting `$PKG_CONFIG`, when set). This works for my cross-compilation workflow, and I suspect one could also set this environment variable for OPAM-cross builds.

What do you think?

**Comment by toots** [comment]
In practice, `PKG_CONFIG_PATH` is usually set to the path to the cross-compilation pkg-config files. The path to `pkg-config` is not usually changed.

At the moment, all currently existing examples of cross-compiled modules do have the `PKG_CONFIG_PATH` variable set unconditionally to the cross-compilation files in the `opam-cross` CI tests.

However, I believe that it would be wiser and more future-proof to allow compilation for both native host and cross-compilation host as `dune` can compile binaries for the native host if they are needed during the build, e.g. to generate some files.

This scenario does currently happen but not (yet?) in the context of `pkg-config`-related stuff.

For this reason, the ability to add a `PKG_CONFIG_PATH` environment variable (and more generally any environment variable) inside the toolchain definition would be the best solution. Once the environment variable is set, the rest of the build can continue transparently.

Lastly, and also out of scope for this, I wanted to point out that it would also be great to be able to tell `dune` to build a binary for a specific host explicitly. This could be used to extract internal data specific to the target host, for instance offset in C structures. `pkg-config` could also be used there for cross-compiling.

**Comment by anmonteiro** [comment]
> In practice, PKG_CONFIG_PATH is usually set to the path to the cross-compilation pkg-config files. The path to pkg-config is not usually changed.

I think we're perhaps talking about 2 different features. Your original comment mentioned dune couldn't find `x86_64-w64-mingw32-pkg-config`. My PR fixes this exact issue.

> For this reason, the ability to add a PKG_CONFIG_PATH environment variable (and more generally any environment variable) inside the toolchain definition would be the best solution. Once the environment variable is set, the rest of the build can continue transparently.

Setting environment variables in findlib.conf seems possible if I'm reading this snippet correctly:

https://github.com/ocaml/dune/blob/adcdd090b4360bef63e7f8eb770b7b33cecb80bc/src/dune_rules/findlib/findlib.ml#L135-L138

There doesn't seem to be a way to set it in the host and the cross-toolchain at the same time, though.

**Comment by toots** [comment]
> > In practice, PKG_CONFIG_PATH is usually set to the path to the cross-compilation pkg-config files. The path to pkg-config is not usually changed.
> 
> I think we're perhaps talking about 2 different features. Your original comment mentioned dune couldn't find `x86_64-w64-mingw32-pkg-config`. My PR fixes this exact issue.

Yes! And this would be great thanks!

> > For this reason, the ability to add a PKG_CONFIG_PATH environment variable (and more generally any environment variable) inside the toolchain definition would be the best solution. Once the environment variable is set, the rest of the build can continue transparently.
> 
> Setting environment variables in findlib.conf seems possible if I'm reading this snippet correctly:
> 
> https://github.com/ocaml/dune/blob/adcdd090b4360bef63e7f8eb770b7b33cecb80bc/src/dune_rules/findlib/findlib.ml#L135-L138
> 
> There doesn't seem to be a way to set it in the host and the cross-toolchain at the same time, though.

Very interesting. Yeah, if env could be filtered by toolchain, i.e.:

```
PKG_CONFIG_PATH(env,windows) = "/path/to/files"
```
That would probably be enough to make it work.

**Comment by anmonteiro** [comment]
> Very interesting. Yeah, if env could be filtered by toolchain, i.e.:
> 
> ```
> PKG_CONFIG_PATH(env,windows) = "/path/to/files"
> ```
> 
> That would probably be enough to make it work.

I think this is possible after #7428. I'll add a test for this after we merge it to check whether it's possible.

 I'll leave this issue open even after we merge #7469 until we actually fix the environment use case.
